db/silkworm/db/chaindb.cpp

```diff
@@ -456,23 +457,60 @@ std::optional<ByteView> Table::get(ByteView key) {
 }
 
 std::optional<ByteView> Table::get(ByteView key, ByteView sub_key) {
-    MDB_val key_val{db::to_mdb_val(key)};
-    MDB_val data{db::to_mdb_val(sub_key)};
-    int rc{get(&key_val, &data, MDB_GET_BOTH_RANGE)};
+    MDB_val mdb_key{db::to_mdb_val(key)};
+    MDB_val mdb_val{db::to_mdb_val(sub_key)};
+    const int rc{get(&mdb_key, &mdb_val, MDB_GET_BOTH_RANGE)};
     if (rc == MDB_NOTFOUND) {
-        return {};
+        return std::nullopt;
     }
     err_handler(rc);
 
-    ByteView x{db::from_mdb_val(data)};
+    ByteView x{db::from_mdb_val(mdb_val)};
     if (!has_prefix(x, sub_key)) {
-        return {};
+        return std::nullopt;
     } else {
         x.remove_prefix(sub_key.length());
         return x;
     }
 }
 
+std::optional<ByteView> Table::seek_dup(ByteView key, ByteView lower_bound) {
+    assert(!lower_bound.empty());
+    MDB_val mdb_key{db::to_mdb_val(key)};
+    MDB_val mdb_val{db::to_mdb_val(lower_bound)};
+    const int rc{get(&mdb_key, &mdb_val, MDB_GET_BOTH_RANGE)};
+    if (rc == MDB_NOTFOUND) {
+        return std::nullopt;
+    }
+    err_handler(rc);
+    return db::from_mdb_val(mdb_val);
+}
+
+std::optional<db::Entry> Table::get_next() {
+    MDB_val mdb_key;
+    MDB_val mdb_val;
+    const int rc{get(&mdb_key, &mdb_val, MDB_NEXT)};
+    if (rc == MDB_NOTFOUND) {
+        return std::nullopt;
+    }
+    err_handler(rc);
+    db::Entry out;
+    out.key = db::from_mdb_val(mdb_key);
+    out.value = db::from_mdb_val(mdb_val);
+    return out;
+}
+
+std::optional<ByteView> Table::get_next_dup() {
```

Comment:
NextDup always returns same key which you passed as input parameter (pointer to same memory region). It means maybe you don’t need return it to App.